File: info/1. Алгоритмы/2. Сортировка/18. Быстрая сортировка. Оптимизация/code sample/SampleLomuto2.java
package sample1;

import java.util.Arrays;

public class SampleLomuto2 {

    public static void main(String[] args) {
        int[] array = new int[] { 0, 5, -2, 7, 3 };
        quickSort(array);
        System.out.println(Arrays.toString(array));
    }

    public static void quickSort(int[] array) {
        quickSort(array, 0, array.length - 1);
    }

    public static void quickSort(int[] array, int lo, int hi) {
        if (hi - lo <= 32) {
            insertionSort(array, lo, hi);
            return;
        }
        int h = breakPartition(array, lo, hi);
        quickSort(array, lo, h - 1);
        quickSort(array, h + 1, hi);
    }

    public static int breakPartition(int[] array, int lo, int hi) {
        int j = lo;
        int supportElement = array[lo];
        for (int i = lo + 1; i <= hi; i++) {
            if (array[i] < supportElement) {
                j+=1;
                swap(array, i, j);
            }
        }
        swap(array, lo, j);
        return j;
    }

    public static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    public static void insertionSort(int[] array, int begin, int end) {
        for (int i = begin; i <= end; i++) {
            int pasteElement = array[i];
            int j;
            for (j = i; j > begin; j--) {
                if (array[j - 1] <= pasteElement) {
                    break;
                }
                array[j] = array[j - 1];
            }
            array[j] = pasteElement;
        }
    }

}

File: info/1. Алгоритмы/2. Сортировка/18. Быстрая сортировка. Оптимизация/code sample/sampleLomuto2.py
def insertion_sort(sequince, start_index, end_index):
    for i in range(start_index, end_index+1):
        paste_element = sequince[i]
        while i > start_index and sequince[i-1] > paste_element:
            sequince[i] = sequince[i-1]
            i = i-1
        sequince[i] = paste_element


def quick_sort(sequince, lo_index=None, hi_index=None):
    if lo_index is None:
        lo_index = 0
    if hi_index is None:
        hi_index = len(sequince)-1
    if hi_index-lo_index <= 32:
        insertion_sort(sequince, lo_index, hi_index)
        return None
    h = partition(sequince, lo_index, hi_index)
    quick_sort(sequince, lo_index, h-1)
    quick_sort(sequince, h+1, hi_index)


def partition(sequince, lo_index, hi_index):
    support_element = sequince[lo_index]
    j = lo_index
    for i in range(lo_index + 1, hi_index + 1):
        if sequince[i] < support_element:
            j += 1
            sequince[i], sequince[j] = sequince[j], sequince[i]
    sequince[lo_index], sequince[j] = sequince[j], sequince[lo_index]
    return j


sequince = [0, 5, -2, 7, 3]

quick_sort(sequince)

print(sequince)
